zebra-state/src/service/read/difficulty.rs

```diff
@@ -0,0 +1,64 @@
+//! Get context and calculate difficulty for the next block.
+
+use std::borrow::Borrow;
+
+use zebra_chain::{
+    block::{Block, Hash, Height},
+    parameters::{Network, POW_AVERAGING_WINDOW},
+    work::difficulty::CompactDifficulty,
+};
+
+use crate::service::{
+    any_ancestor_blocks,
+    check::{difficulty::POW_MEDIAN_BLOCK_SPAN, AdjustedDifficulty},
+    finalized_state::ZebraDb,
+    NonFinalizedState,
+};
+
+/// Return the CompactDifficulty for the current best chain.
+///
+/// Return `None` if we don't have enough blocks in the state.
+pub fn relevant_chain_difficulty(
+    non_finalized_state: &NonFinalizedState,
+    db: &ZebraDb,
+    tip: (Height, Hash),
+    network: Network,
+) -> Option<CompactDifficulty> {
+    let relevant_chain = any_ancestor_blocks(non_finalized_state, db, tip.1);
+    difficulty(relevant_chain, tip.0, network)
+}
+
+fn difficulty<C>(
+    relevant_chain: C,
+    tip_height: Height,
+    network: Network,
+) -> Option<CompactDifficulty>
+where
+    C: IntoIterator,
+    C::Item: Borrow<Block>,
+    C::IntoIter: ExactSizeIterator,
+{
+    const MAX_CONTEXT_BLOCKS: usize = POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN;
+
+    let relevant_chain: Vec<_> = relevant_chain
+        .into_iter()
+        .take(MAX_CONTEXT_BLOCKS)
+        .collect();
+
+    let relevant_data = relevant_chain.iter().map(|block| {
+        (
+            block.borrow().header.difficulty_threshold,
+            block.borrow().header.time,
+        )
+    });
+
+    if relevant_data.len() < MAX_CONTEXT_BLOCKS {
+        return None;
+    }
+
+    let time = chrono::Utc::now();
```

Comment:
This time is used to check if the next block will be a testnet minimum difficulty block. So it needs to be exactly the same time as we use in the `getblocktemplate` RPC response.

We can't call `chrono::Utc::now()` twice, because we might get different times.

Here is one possible design:
1. If we're on testnet, and minimum difficulty blocks are allowed by any times in the valid range, set the minimum time to the lowest time that allows minimum difficulty blocks (this avoids the difficulty changing if the miner changes the time)
2. Clamp the time using the minimum and maximum time consensus rules
3. Return the minimum time, and the time we used to calculate the difficulty, as part of the state response

Reply:
You added like 5 comment about time ? Can you just take the important information and place it in a single comment ? It is hard to follow.